### ISSUE TITLE
Validation done after Payment details accepted

### ISSUE THREAD

**Issue by mikkamp** [woocommerce/woocommerce-gateway-stripe#161]
When Stripe is set to use a modal "Enable Stripe Checkout" the payment information is filled out in a modal form after clicking on the "Continue to payment" button.

The user then enters their payment details and if these are correct it shows a green checkbox (which visually suggests the payment has been accepted).

![Checkout](https://cloudup.com/c-Oafz_D9SV+)
Checkout image link: https://cloudup.com/c-Oafz_D9SV

However the actual validation is only done after this, which means an error can be received. In my test case I generated an error by not selecting a shipping method which returns the error "No shipping method has been selected. Please double check your address, or contact us if you need any help". But there could be many reasons why an order is not validated.

At this point the user must correct the error (if possible) and then enter their payment details again. Because of the way it shows the green checkbox a second time it looks like the payment was completed a second time. The payment is handled correctly and the user is only actually charged if the order is successfully processed without any errors.

However to the user this is confusing. A pre-validation should be done before the payment details are entered, or alternatively they should be saved so after the user corrects the error they do not have to pass through the payment process again.

Affected ticket: 525547-zd-woothemes

**Comment by ApplicationError** [comment]
I'm pretty sure I just filed a duplicate of this. I just encountered this issue as well.
Who do I need to pay how much to fix this?

**Comment by ApplicationError** [comment]
Thanks @roykho. Is there any way I can help speed this fix up?

**Comment by roykho** [comment]
This seems to happen only when shipping is involved I believe. I tested the other form fields and indeed it does validate those prior to the payment form popping up. If you have more to add, let me know. Meanwhile I will look into shipping.

**Comment by ApplicationError** [comment]
Hi @roykho I get the error and we have no shipping information being collected or used. I could share a link in private.

Different note, if you know a really good WooCommerce/WP dev for hire - PLEASE let me know.

**Comment by roykho** [comment]
@ApplicationError - then you might have other issues. Try switching to a WP default theme like 2017 and re-test.

**Comment by ApplicationError** [comment]
Um. So we closed my ticket for no reason then? And what would have theme have to do with the form validation?

**Comment by roykho** [comment]
@ApplicationError some themes are not following the way WC does things and can cause this type of issue. Not saying it is but worth a test.

**Comment by ApplicationError** [comment]
@roykho I don't have any other themes installed and I don't have another environment to test this on. I'm also not keen to install anything that we won't be using. I chatted with Avada about it, and they said it's they are not touching that functionality. I also checked the woo-config.php that Avada uses, and no issues there.

**Comment by roykho** [comment]
@ApplicationError Unfortunately I won't be able to help there. You need to setup a staging environment to test things out such as this. Its a common practice for store owners.

**Comment by ApplicationError** [comment]
I am working on our staging environment, it is not in production yet. 

Just a note, if I use PayPal as an option, the form field validation catches the wrong email address (formatting) as soon as I click "Proceed to PayPal". This seems to support the assumption that it's a plugin/extension bug.

**Comment by roykho** [comment]
Its possible PayPal is checking for a valid email where Stripe is checking if the field is empty.

**Comment by ApplicationError** [comment]
I contacted Stripe and they said it was an issue with the extension. I also contacted WooCommerce, they also said it was likely an issue with the extension. So should I take this back to Stripe as a bug on their side?

**Comment by roykho** [comment]
No, Stripe as in this plugin you're using. Not Stripe the company.

**Comment by ApplicationError** [comment]
Got it. 10-4. I don't really know what I'm doing, but I'll try to poke around in the code and see if I can compare it to the PayPal extension. See if it does checks differently.

**Comment by roykho** [comment]
You can test this yourself by emptying the email field and click on "proceed to checkout". I am guessing you will see validation of some sort before the popup.

**Comment by ApplicationError** [comment]
If there is no email (field blank) then: Billing Email Address is a required field.
If the there is something in the email field (and: Use a new payment method) then: Stripe window/modal is shown -> followed by error after submitting payment details (Email Address is not a valid email address.)

**Comment by roykho** [comment]
Yeah like I suspected, it is only validating if the fields are empty. Not if the email field has a correctly formatted address. I have to look into it.

**Comment by ApplicationError** [comment]
You're a rock star. Thank you!

**Comment by roykho** [comment]
@ApplicationError - can you please download from the master branch and see if it fixes your email format issue?

**Comment by ApplicationError** [comment]
Hi @roykho sorry for the delay. Works. WooCommerce shows the field validation error and does not show the modal until the error is resolved.

**Comment by ApplicationError** [comment]
I found a "related" issue. If the email address of the customer is already associated with a customer account an error message is thrown "An account is already registered with your email address. Please login." -- Right now this error message is shown AFTER going through the Stripe payment. 

The PayPal option detects the problem before redirecting to PayPal for payment.

**Comment by roykho** [comment]
@ApplicationError can you show a screenshot of all your settings for checkout/account so I can try and replicate this?

**Comment by ApplicationError** [comment]
Here we go... 

![screencapture-alpha-checkout](https://cloud.githubusercontent.com/assets/26173192/24280407/ec328b74-1015-11e7-949d-3da37ae92996.png)

![screencapture-alpha-account](https://cloud.githubusercontent.com/assets/26173192/24280438/2bf97b28-1016-11e7-8ef4-f56f79b78bc0.png)

**Comment by ApplicationError** [comment]
Any updates on this?

**Comment by mikestern** [comment]
Is there a fix for this issue yet? It's a terrible customer experience as-is.

**Comment by ApplicationError** [comment]
@mikestern We've given up on getting this fixed here, pretty much. We'll create our own version in the coming weeks and make sure these things are fixed.

**Comment by roykho** [comment]
@mikkamp - so circling back to this. I like the idea of having the token saved so even validation error happens, we don't need to show the modal a second time. However this will also mean the checkout form can later be changed before submitting again. Are we ok with that?

**Comment by Frique** [comment]
@roykho No, that could lead to potential security issues, conflict and confusion. Why can't the modal be triggered only after all form validation is done? Similar to redirecting to PayPal only after validation is done.

**Comment by roykho** [comment]
@Frique because this is not PayPal. There is no redirect. If you guys are so concern with this why not just use Stripe default checkout form instead of the modal??

**Comment by Frique** [comment]
@roykho Sorry, just trying to understand since I don't see why it's such a complicated issue yet. To my understanding WooCommerce handles the form validation during checkout (via ajax) and then returns a status - either checkout errors or a pass. The payment gateway (whichever) waits for this success event to trigger their next step which can be a redirect for PayPal and a modal popup for Stripe?

"why not just use Stripe default checkout form instead of the modal?" Because of better UX and conversion rates when using the modal.

**Comment by roykho** [comment]
@Frique do you want to submit a PR?

**Comment by Frique** [comment]
@roykho Unlikely since I don't fully understand the issue but I guess I'll take a look sometime.

**Comment by roykho** [comment]
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/tree/4.0.0

**Comment by WebAssembler** [comment]
Hi there @roykho thanks for fixing this issue in version 4.0.0

I really need to implement the fix asap but I'm wary of deploying the beta version on a production website. Would there be any way the fix could be merged into the stable master branch prior to the v4 release?

**Comment by Flo-Ru** [comment]
Hi,
fixed our problem of the validation AFTER the payment popup with the 4.0.0 beta version. 
It is now working on the staging site. 
WHEN will there be an update to the WordPress version?
I am a little concerned to use the beta on the production site...
Thx.

**Comment by dotbloup** [comment]
I also faced this issue today.

**Comment by roykho** [comment]
Unfortunately this is a bad decision on my part to fix this via a pre-validation. Sure it fixed it for all you nice folks that had this issue, but now it can possibly break other people's sites if they have custom fields that are required.

What the fix did was simply call the same checkout validation via AJAX before the popup shows and it will take every required field on the checkout and validate it.

We need to think of another way to resolve this. Perhaps just a simple client side validation.

**Comment by roykho** [comment]
Continue with #446

**Comment by roykho** [comment]
I have removed the validation from checkout for NON Stripe Modal checkouts. This means it is back to how it used to function pre 4.0.0. I would suggest you give this branch a test and see if it will work with your current setup https://github.com/woocommerce/woocommerce-gateway-stripe/tree/validation

PS: Stripe Modal checkout will still get pre validation.

**Comment by anoordende** [comment]
Using 4.0.7 and certain that Stripe Modal checkout is NOT pre-validating.

**Comment by roykho** [comment]
@anoordende that is correct it does not. We later decided not to do it anymore as it caused sites to break.
